### PR TITLE
add cluster information test cases of REST service

### PR DESCRIPTION
**Pull request by skw1992**

### CODE REVIEW

File: integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala [hunk 1]
```diff
@@ -108,18 +109,54 @@ class RestServiceSpec extends TestSpecBase {
 
   "cluster information" should {
     "retrieve 1 master for a non-HA cluster" in {
+      // setup
+      val expectMastersCount = cluster.getMasterHosts.size
+      val expectMasterHost = cluster.getMasterHosts.head
 
+      // exercise
+      restClient.listMasters().cluster.size shouldEqual expectMastersCount
+      restClient.listMasters().cluster.head._1 shouldEqual expectMasterHost
```
Comment (stanleyxu2005) [inline integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala]: Not only check number. Should also verify running status, hostname etc.

File: integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala [hunk 2]
```diff
@@ -108,18 +109,54 @@ class RestServiceSpec extends TestSpecBase {
...
     }
 
     "retrieve 0 worker, if cluster is started without any workers" in {
       // todo: defect
+      // setup
+      val oldCluster = cluster
+      oldCluster.shutDown()
+      val newCluster = new MiniCluster(workerNum = 0)
```
Comment (stanleyxu2005) [inline integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala]: Try with `killWorkerNode` until worker number == 0. At cleanup stage, add them back.

File: integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala [hunk 3]
```diff
@@ -108,18 +109,54 @@ class RestServiceSpec extends TestSpecBase {
...
+      newCluster.start()
+
+      // exercise
+      try {
+        newCluster.restClient.listWorkers().size shouldEqual 0
+      } finally {
+        newCluster.shutDown()
+        oldCluster.start()
+        Thread.sleep(5000)
```
Comment (stanleyxu2005) [inline integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala]: Dont sleep with a magic number. Try something like `retryUtil(restClient.getWorker(0)!= null)`

File: integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala [hunk 4]
```diff
@@ -108,18 +109,54 @@ class RestServiceSpec extends TestSpecBase {
...
+      }
     }
 
     "retrieve the same number of workers as cluster has" in {
+      // setup
+      val expectWorkersCount = cluster.getWorkerHosts.size
+
+      // exercise
+      restClient.listWorkers().size shouldEqual expectWorkersCount
```
Comment (stanleyxu2005) [inline integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala]: Also verify the hostname, running status etc.

File: integrationtest/src/main/scala/io/gearpump/integrationtest/minicluster/RestClient.scala [hunk 5]
```diff
@@ -44,6 +47,20 @@ class RestClient(host: String, port: Int) {
     callFromRoot("version")
   }
 
+  def listMasters(): MasterSummary = try {
```
Comment (stanleyxu2005) [inline integrationtest/src/main/scala/io/gearpump/integrationtest/minicluster/RestClient.scala]: Masters should be a List with one element. What the method returns is `getMasterLeader`

**Comment by skw1992** [comment]
@stanleyxu2005 updated, please review

File: integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala [hunk 6]
```diff
@@ -142,18 +143,64 @@ class RestServiceSpec extends TestSpecBase {
 
   "cluster information" should {
     "retrieve 1 master for a non-HA cluster" in {
+      // setup
+      val expectMastersCount = cluster.getMasterHosts.size
```
Comment (stanleyxu2005) [inline integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala]: Should use `expected` not `expect` for a variable. 

`expecto` is used for method, which means the method will expect particular behavior. If the behavior is unexpected, an exception will be thrown.

File: integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala [hunk 7]
```diff
@@ -142,18 +143,64 @@ class RestServiceSpec extends TestSpecBase {
...
+      val expectMaster = cluster.getMasters.head
 
-    }
-
-    "retrieve 0 worker, if cluster is started without any workers" in {
-      // todo: defect
+      // exercise
+      restClient.listMasters().size shouldEqual expectMastersCount
```
Comment (stanleyxu2005) [inline integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala]: Suggestion: store the result of `listMasters()` as a variable. Do not call the methods at every verification step.

File: integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala [hunk 8]
```diff
@@ -142,18 +143,64 @@ class RestServiceSpec extends TestSpecBase {
...
+      restClient.listRunningWorkers().size shouldEqual expectWorkersCount
+      restClient.listRunningWorkers().foreach { worker =>
+        worker.state shouldEqual MasterToAppMaster.AppMasterActive
+      }
     }
 
     "find a newly added worker instance" in {
+      // setup
+      val oldWorkersCount = restClient.listRunningWorkers().size
+      val workerName = s"newWorker"
```
Comment (stanleyxu2005) [inline integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala]: The modifier `s` is not necessary. `s"newWorker"` could be `"newWorker". Please check IDE warnings before you commit code. We have a great IntelliJ to find warnings for you. 

Not an obvious issue but how can you be sure, there is no existing worker called "newWorker"?

File: integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala [hunk 9]
```diff
@@ -142,18 +143,64 @@ class RestServiceSpec extends TestSpecBase {
...
+        Util.retryUntil(restClient.listRunningWorkers().size > oldWorkersCount)
+        restClient.listRunningWorkers().size shouldEqual (oldWorkersCount + 1)
+      } finally {
+        cluster.removeWorkerNode(workerName)
+        Util.retryUntil(restClient.listRunningWorkers().size == oldWorkersCount)
+      }
+    }
+
+    "retrieve 0 worker, if cluster is started without any workers" in {
+      // todo: defect
```
Comment (stanleyxu2005) [inline integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala]: Remove `//todo`!

File: integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala [hunk 10]
```diff
@@ -142,18 +143,64 @@ class RestServiceSpec extends TestSpecBase {
...
+      // setup
+      val originWorkersCount = cluster.getWorkerHosts.size
+
+      // exercise
+      try {
+        restClient.listRunningWorkers().size shouldEqual originWorkersCount
+        cluster.getWorkerHosts.foreach { host =>
```
Comment (stanleyxu2005) [inline integrationtest/src/it/scala/io/gearpump/integrationtest/checklist/RestServiceSpec.scala]: `foreach` is wrong! Use `while size >0`

File: integrationtest/src/main/scala/io/gearpump/integrationtest/minicluster/MiniCluster.scala [hunk 11]
```diff
@@ -125,6 +125,10 @@ class MiniCluster(
     }
   }
 
+  def getMasters = {
```
Comment (stanleyxu2005) [inline integrationtest/src/main/scala/io/gearpump/integrationtest/minicluster/MiniCluster.scala]: we have already `def getMasterHosts = {`. Why add this new method?!

File: integrationtest/src/main/scala/io/gearpump/integrationtest/minicluster/RestClient.scala [hunk 12]
```diff
@@ -45,6 +48,28 @@ class RestClient(host: String, port: Int) {
...
 
+  def listMasters(): List[(String, Int)] = try {
+    queryMaster().cluster
+  }
+
+  def queryMaster(): MasterSummary = try {
+    val resp = callApi("master")
+    read[MasterData](resp).masterDescription
+  } catch {
+    case ex: Throwable => null
```
Comment (stanleyxu2005) [inline integrationtest/src/main/scala/io/gearpump/integrationtest/minicluster/RestClient.scala]: No need to wrap with `try`. Let it fail, if the result is not available

File: integrationtest/src/main/scala/io/gearpump/integrationtest/minicluster/RestClient.scala [hunk 13]
```diff
@@ -45,6 +48,28 @@ class RestClient(host: String, port: Int) {
     callFromRoot("version")
   }
 
+  def listMasters(): List[(String, Int)] = try {
```
Comment (stanleyxu2005) [inline integrationtest/src/main/scala/io/gearpump/integrationtest/minicluster/RestClient.scala]: Return List.empty, if failed to get masters

File: integrationtest/src/main/scala/io/gearpump/integrationtest/minicluster/MiniCluster.scala (hunk 11 above)
Comment (skw1992) [inline integrationtest/src/main/scala/io/gearpump/integrationtest/minicluster/MiniCluster.scala]: The `getMasterHosts` method only returns the list of masters' hosts, I also need the ports information of masters to be verified.

**Comment by stanleyxu2005** [comment]
+1

**Pull request merged**